File: HouseRobber_LC198.java
//Time complexity: O(n * 2) => O(n)
//Space complexity: O(2n) =>  o(n)
//n = length of nums

class Solution {
    public int rob(int[] nums) {
        if(nums == null || nums.length == 0)
            return 0;
        
        int m = nums.length;
        int n = 2; // choose and not choose
        
        int[][] dp = new int[m][n];
        //fill the 1st row with initial values in dp
        dp[0][0] = nums[0];
        dp[0][1] = 0;
        
        for(int i = 1; i < m; i++){
            for(int j = 0; j < n; j++){
                if(j == 0){
                    int prevDC = dp[i-1][1];
                    dp[i][0] = nums[i] + prevDC;
                }
                else{
                    int prevC = dp[i-1][0];
                    int prevDC = dp[i-1][1];
                    
                    dp[i][1] = Math.max(prevC, prevDC);
                }
            }
        }
        
        //return max of choose, not choose of last row
        return Math.max(dp[m -1][0], dp[m -1][1]);
    }
}

File: houseRobber-LC 198.py
# Time Complexity = O(n)
# Space Complexity = O(1), since we are storing values in variables

class Solution:
    def rob(self, nums: list[int]) -> int:
        if len(nums) == 0:
            return 0
        
        # Initialise two variables, one with 0, and the other with the first element of array
        # Doing this method because our next result only depends upon the previous 2 results and not the complete array
        skip = 0
        take = nums[0]
        
        # Iterating from 1, since 0th element already initialised above
        for i in range(1, len(nums)):
            temp = skip                 # Storing the value of skip, as that will be used for the    next iteration
            skip = max(skip, take)
            take = temp + nums[i]
        
        return max(skip, take)

        